### PR TITLE
Dynamo Query API: Support KeyConditionExpression

### PR DESCRIPTION
**Pull request by li-boxuan**
**What this PR does**:
1. Add the skeleton for Query API
2. Support [KeyConditionExpression](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#DDB-Query-request-KeyConditionExpression) (the most important parameter in Query API) 

**What this PR does not address**:
1. Exactly the same error message as DynamoDB
3. Legacy parameters
4. FilterExpression
5. ProjectionExpression
6. Limit
7. Pagination
8. ConsistentRead

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+   *
+   * <p>Examples:<br>
+   * partitionKeyName = :partitionkeyval AND sortKeyName = :sortkeyval<br>
+   * partitionKeyName = :partitionkeyval AND sortKeyName < :sortkeyval<br>
+   * partitionKeyName = :partitionkeyval AND sortKeyName <= :sortkeyval<br>
+   * partitionKeyName = :partitionkeyval AND sortKeyName > :sortkeyval<br>
+   * partitionKeyName = :partitionkeyval AND sortKeyName >= :sortkeyval<br>
+   */
+  private static final Pattern OP_COMPARISON_PATTERN =
+      Pattern.compile("(\\S+)\\s*=\\s*(:\\S+)\\s*AND\\s*(\\S+)\\s*([<>=]+)\\s*(:\\S+)");
```
Comment (li-boxuan) [inline sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java:55]: We should split this pattern into two parts. First, we match the equality pattern, and then we match the operator comparison pattern for the unmatched substring. By doing so, we could reuse the regex and match logic in FilterExpression.

**Comment by li-boxuan** [comment]
Right now, we are using Regex to match KeyConditionExpression. It is probably the fastest way (needs benchmarking) but less extensible, flexible, and readable compared to an Antlr (parser) approach. Later, we could revisit this part and refactor using Antlr if needed.

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      Pattern.compile(
+          "(?i)(\\S+)\\s*=\\s*(:\\S+)\\s*AND\\s*begins_with\\s*\\(\\s*(\\S+),\\s*(:\\S+)\\s*\\)");
+
+  public QueryResult query(QueryRequest queryRequest, StargateBridgeClient bridge) {
+    // Get table schema
+    final String tableName = queryRequest.getTableName();
+    Schema.CqlTable table =
+        bridge
+            .getTable(KEYSPACE_NAME, tableName)
+            .orElseThrow(() -> new IllegalArgumentException("Table not found"));
```
Comment (tatu-at-datastax) [inline sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java]: Maybe include table name in the error message?

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    QueryOuterClass.ColumnSpec partitionKey = table.getPartitionKeyColumns(0);
+    List<QueryOuterClass.ColumnSpec> clusteringKeys = table.getClusteringKeyColumnsList();
+    Optional<QueryOuterClass.ColumnSpec> clusteringKey = Optional.empty();
+    if (CollectionUtils.isNotEmpty(clusteringKeys)) {
+      assert clusteringKeys.size() == 1;
```
Comment (tatu-at-datastax) [inline sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java:88]: If this needs actual validation better check and throw regular exception? (if it should never occur, assert fine)

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    if (beginWithMatcher.find()) {
+      // TODO: address this
+      throw new IllegalArgumentException("Cassandra does not support begin_with query");
+    }
+
+    throw new IllegalArgumentException(keyConditionExpression + " is not a valid expression");
+  }
+
+  private String getKeyName(String rawKeyName, Map<String, String> attributeNames) {
+    if (rawKeyName.charAt(0) == '#') {
```
Comment (tatu-at-datastax) [inline sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java:237]: May want to document why this is being checked?

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/resources/DynamoResource.java [hunk 5]
```diff
@@ -85,17 +92,24 @@ public Response handleRequest(
...
+          QueryRequest queryRequest = awsRequestMapper.readValue(payload, QueryRequest.class);
+          result = queryProxy.query(queryRequest, bridge);
+          break;
         default:
           throw new WebApplicationException(
               "Invalid statement type: " + target, Response.Status.BAD_REQUEST);
       }
       response = awsRequestMapper.writeValueAsBytes(result);
     } catch (JsonProcessingException ex) {
+      logger.info("Invalid payload", ex);
```
Comment (tatu-at-datastax) [inline sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/resources/DynamoResource.java]: Typically one should either throw an exception further OR log, but not both? In this case, specifically, logs could be full of "Invalid payload" with no useful information (request id, caller, etc etc). Then again, exception message itself does not have much more information.

Of course there may be value in logging something about invalid requests beyond metrics so sometimes logging makes sense too.

**Review by tatu-at-datastax — DISMISSED** [review]

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java (hunk 2 above)
Comment (li-boxuan) [reply sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java]: Sure thing. I created https://github.com/li-boxuan/stargate/issues/19 to track all the exception handling. Later, we need to make sure the error messages align with DynamoDB's error messages.

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java (hunk 3 above)
Comment (li-boxuan) [reply sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java:88]: It never occurs, unless our code is buggy (When creating the table, we ensure there is only one clustering key). This is more of a hint for developers.

File: sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java (hunk 4 above)
Comment (li-boxuan) [reply sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/dynamo/QueryProxy.java:237]: Sure, I will add some comments. To simply put, a word starting with a pound sign is a placeholder, whose real value needs to be fetched from `ExpressionAttributeNames`.

ref: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ExpressionAttributeNames.html

**Pull request merged**